### PR TITLE
settings,settingswatcher: repair a regression

### LINKED ISSUE #111539: roachtest: change-replicas/mixed-version failed
roachtest.change-replicas/mixed-version [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/11994276?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/11994276?buildTab=artifacts#/change-replicas/mixed-version) on master @ [fad649d89721ddb3e9f3dcab1ad5d14f74c91bc9](https://github.com/cockroachdb/cockroach/commits/fad649d89721ddb3e9f3dcab1ad5d14f74c91bc9):

```
(test_runner.go:1166).runTest: test timed out (20m0s)
(versionupgrade.go:345).func1: timed out after 10m0s: expected n1 to be at cluster version 23.1-26, but is still at 23.1: read tcp 172.17.0.3:53846 -> 34.138.21.241:26257: read: connection reset by peer
test artifacts and logs in: /artifacts/change-replicas/mixed-version/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-11994276/change-replicas-mixed-version/1696052742802/1696054092661)

</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*change-replicas/mixed-version.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-31955

### LINKED ISSUE #111610: settings: (regression) stale values coming from local cache
**Describe the problem**

Some time recently (in the past week) two groups of tests have started to fail:

- this roachtest failure: #111539

  ```
  server/auto_upgrade.go:101 ⋮ [T1,Vsystem,n3] 119  error when finalizing cluster version upgrade: ‹set-version›: cannot upgrade to 23.1-26: cluster.preserve _downgrade_option is set to ‹23.1›
  ```

  which looks like a new occurrence of #68335.

- CI after rebase for this PR from @healthy-pod #102427
 
  ```
  testcluster.go:495: set-cluster-version: node at 1000023.1-26 cannot run 1000023.1 (minimum version is 1000023.1-26)
  ```

They seem related, and also related to this PR: #110676.

Epic: CRDB-6671

Jira issue: CRDB-31987

### PR DESCRIPTION
**Pull request by knz**
Many thanks to @RaduBerinde for helping identify the problem.

Fixes #111610. 
Fixes #111539.
Epic: CRDB-6671

TLDR: this commit fixes a regression introduced in
https://github.com/cockroachdb/cockroach/commit/2f5d717178a87b42fa94c20a226cc491d185455d by undoing part of that
change and fixing the problem that change was aiming to solve in a
different way.

Context: there is a method on the "setting updater" object (that
returned by `settings.MakeUpdater`) called `ResetRemaining()`. Its
purpose is to "reset all settings not modified earlier, to default
values'.

Prior to the commit linked above, the definition for
`ResetRemaining()` was "reset all settings not modified
earlier *through the same updater*, to default values". Then the
linked commit changed that to "reset all settings not modified
earlier *in any way*, to default values".

The motivation for this earlier commit boiled down to the need to
preserve customizations set by the test-only `.Override()` method. We
wanted to ensure that the value modified through `.Override()` was
preserved by `.ResetRemaining()` even though the `.Override()` method
does not take an updater as argument.

Changing the definition of `.ResetRemaining()` in that way, however,
introduced a regression in a separate area: it made it impossible for
the settings watcher (reading values from `system.overrides`) to clear
any stale values loaded earlier from the local persistent cache. Since
the load from persistent cache was also "modifying a value", the
`ResetRemaining()` call in the settings watcher was preserving stale
values. This was a new bug/regression, and was detected through
various test failures.

To address this bug, it is thus necessary to restore the original
definition of `ResetRemaining()`, which the present patch does through
a partial revert of https://github.com/cockroachdb/cockroach/commit/2f5d717178a87b42fa94c20a226cc491d185455d.

To ensure that the `.Override()` customizations remain preserved, the
patch makes `ResetRemaining()` look at the (new)value origin to ensure
setting values modified with origin `OriginOverride` are preserved.

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/settings/setting.go` line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrfwKBAPkq-diAJQhk:-NfrfwKBAPkq-diAJQhl:b-zbdkzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf54cd63024496505ee5513ae398bf5d24002c5a/pkg/settings/setting.go#L304)):*
> ```Go
> 	// via a host-cluster override for this or all tenant(s).
> 	OriginExternallySet
> 	// OriginOverride indicates the value has been set via an test override.
> ```

[nit] a test
___
*[`pkg/settings/updater.go` line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrgMAaDV5lEdOJu9kH:-NfrgMAb-9MfSfxo8p8B:b-gmqceo) ([raw file](https://github.com/cockroachdb/cockroach/blob/29322216684f0d4aed0bf46a7bf3cb1bdd562691/pkg/settings/updater.go#L250)):*
> ```Go
> 		if _, setInThisUpdater := u.m[v.InternalKey()]; !setInThisUpdater &&
> 			// We need to preserve test overrides.
> 			u.sv.getValueOrigin(ctx, v.getSlot()) != OriginOverride {
> ```

Isn't this racy?

Well, the whole `setValueOrigin` infrastructure feels racy - if two different contexts are trying to set different values and origins, they could interleave their updates and we end up with the value from one and the origin from the other.

We could do all updates under a big mutex (since updates are not perf sensitive) and use atomic reads for reading the current value. We'd need to do this entire "if origin != override { set new value and origin }" operation under the mutex.

**Comment by RaduBerinde** [comment]
*[`pkg/settings/updater.go` line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrgMAaDV5lEdOJu9kH:-NfrjAFaDzweS5DuLJGp:b-bmbfug) ([raw file](https://github.com/cockroachdb/cockroach/blob/29322216684f0d4aed0bf46a7bf3cb1bdd562691/pkg/settings/updater.go#L250)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Isn't this racy?

Well, the whole `setValueOrigin` infrastructure feels racy - if two different contexts are trying to set different values and origins, they could interleave their updates and we end up with the value from one and the origin from the other.

We could do all updates under a big mutex (since updates are not perf sensitive) and use atomic reads for reading the current value. We'd need to do this entire "if origin != override { set new value and origin }" operation under the mutex.
</blockquote></details>

(btw this is a small window and is related to testing only, I have no problem merging this for now and filing an issue to follow up)

**Comment by knz** [comment]
> Well, the whole setValueOrigin infrastructure feels racy - if two different contexts are trying to set different values and origins, they could interleave their updates and we end up with the value from one and the origin from the other.

This is indeed only relevant for testing, right? When a test runs an `.Override` call concurrently with a running server?

Withing a server there's no concurrency, since the load from local cache is guaranteed to have completed by the time the settings watcher starts.

**Comment by RaduBerinde** [comment]
Yeah, I guess we'd need an `.Override` call happening at the same time as us receiving an update on that setting from the table (e.g. another node changed it), shouldn't happen in practice. Would be good to add a comment explaining why it's ok.

**Review by RaduBerinde — APPROVED** [review]

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 8 of 8 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`-- commits` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrtzCW2SO_5uUUKDQ0):*
IIUC the bug was only present when using `Override` which is a test-only facility (which implies that the release note wouldn't be necessary)?
___
*[`pkg/settings/integration_tests/settings_test.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrvErm1pVAPf2JatV2:-NfrvErm1pVAPf2JatV3:b-vn5mc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/29322216684f0d4aed0bf46a7bf3cb1bdd562691/pkg/settings/integration_tests/settings_test.go#L293)):*
> ```Go
> }
> 
> // TestSettingsPersistenceEndToEndDifferentNodew checks that a stale
> ```

nit: "DifferentNodew" isn't present in the test name.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @stevendanna, and @yuzefovich)

___
*[`-- commits` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrtzCW2SO_5uUUKDQ0):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

IIUC the bug was only present when using `Override` which is a test-only facility (which implies that the release note wouldn't be necessary)?
</blockquote></details>

Sadly, no -- the bug appears as a result of my previous commit because it disabled the ability of the settingswatcher to clear the settings not set in system.settings any more but that had a value in the local cache.

You can see this in the regression test I've added - this does not contain .Override calls.
___
*[`pkg/settings/setting.go` line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrfwKBAPkq-diAJQhk:-Nfv7aow43LWOxImIPWX:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf54cd63024496505ee5513ae398bf5d24002c5a/pkg/settings/setting.go#L304)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] a test
</blockquote></details>

Fixed.
___
*[`pkg/settings/updater.go` line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrgMAaDV5lEdOJu9kH:-Nfv7eCh3e3cQU4PAQa4:b-aj5juw) ([raw file](https://github.com/cockroachdb/cockroach/blob/29322216684f0d4aed0bf46a7bf3cb1bdd562691/pkg/settings/updater.go#L250)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

(btw this is a small window and is related to testing only, I have no problem merging this for now and filing an issue to follow up)
</blockquote></details>

I looked further and it appears we're generally safe: .Override calls in test code only ever happen either:
- before a `StartServer` call, when there's no race.
- after `StartServer`, therefore after the local cache is guaranteed to have been loaded already, and the settingswatcher is guaranteed to have completed its initial scan.

So in order to trigger a race we would need one of the following scenarios:
- an async process in a test runs `SET CLUSTER SETTING` concurrently with an .Override call  (we don't do that anywhere currently); AND there's an unexpected rangefeed restart happening at exactly the right time.
- a server for a secondary tenant is started, then a test performs both `[ALTER TENANT] SET CLUSTER SETTING` in system tenant and `.Override` on the secondary  tenant; AND there's an unexpected rangefeed restart happening at exactly the right time.

The combination of factors (especially the rangefeed restart at exactly the right time) seems exceedingly unlikely.

I can file the issue to follow up, but also we could just let this sleep until it happens to become an issue.
___
*[`pkg/settings/integration_tests/settings_test.go` line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111684#-NfrvErm1pVAPf2JatV2:-Nfv953W7i4wApbX8A_r:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/29322216684f0d4aed0bf46a7bf3cb1bdd562691/pkg/settings/integration_tests/settings_test.go#L293)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: "DifferentNodew" isn't present in the test name.
</blockquote></details>

Fixed.

**Review by stevendanna — COMMENTED** [review]
Thanks for handling this. I defer to Yahor and Radu on this one.

**Comment by knz** [comment]
bors r=RaduBerinde,yuzefovich,stevendanna
